src/text/parsers/decimal.rs

```diff
@@ -19,59 +19,53 @@ use crate::types::magnitude::Magnitude;
-                // Needs to return the same fields as above, so we tack on a 'None'
-                map(
-                    pair(opt(tag("-")), digits_before_dot),
-                    |(sign, leading_digits)| (sign, leading_digits, None),
-                ),
-            )),
-            decimal_exponent_marker_followed_by_digits,
-        ),
-        |((sign, digits_before, digits_after), exponent)| {
-            let decimal = decimal_from_text_components(sign, digits_before, digits_after, exponent);
-            TextValue::Decimal(decimal)
-        },
-    )(input)
+fn decimal_with_exponent(input: &str) -> IonParseResult<TextValue> {
+    let (remaining, ((sign, digits_before, digits_after), exponent)) = pair(
+        alt((
+            // Returns a tuple of (sign, digits before '.', and digits after '.')
+            floating_point_number_components,
+            // Needs to return the same fields as above, so we tack on a 'None'
+            map(
+                pair(opt(tag("-")).upgrade(), digits_before_dot),
+                |(sign, leading_digits)| (sign, leading_digits, None),
+            ),
+        )),
+        decimal_exponent_marker_followed_by_digits,
+    )(input)?;
+    let decimal = decimal_from_text_components(sign, digits_before, digits_after, exponent)?.1;
+    Ok((remaining, TextValue::Decimal(decimal)))
 }
 
 /// Matches decimal values that do not have an exponent. (For example, `7.`, `7.1`, and `-7.1`.)
-fn decimal_without_exponent(input: &str) -> IResult<&str, TextValue> {
-    map(
-        floating_point_number_components,
-        |(sign, digits_before_dot, digits_after_dot)| {
-            let decimal = decimal_from_text_components(
-                sign,
-                digits_before_dot,
-                digits_after_dot,
-                "0", // If no exponent is specified, we always start from 0
-            );
-            TextValue::Decimal(decimal)
-        },
-    )(input)
+fn decimal_without_exponent(input: &str) -> IonParseResult<TextValue> {
+    let (remaining, (sign, digits_before_dot, digits_after_dot)) =
+        floating_point_number_components(input)?;
+    let decimal = decimal_from_text_components(
+        sign,
+        digits_before_dot,
+        digits_after_dot,
+        "0", // If no exponent is specified, we always start from 0
+    )?
+    .1;
+    Ok((remaining, TextValue::Decimal(decimal)))
 }
 
 /// Given the four text components of a decimal value (the sign, the digits before the decimal point,
 /// the digits after the decimal point, and the exponent), constructs a [Decimal] value.
-fn decimal_from_text_components(
-    sign_text: Option<&str>,
-    digits_before_dot: &str,
-    digits_after_dot: Option<&str>,
-    exponent_text: &str,
-) -> Decimal {
+fn decimal_from_text_components<'a>(
+    sign_text: Option<&'a str>,
+    digits_before_dot: &'a str,
+    digits_after_dot: Option<&'a str>,
+    exponent_text: &'a str,
+) -> IonParseResult<'a, Decimal> {
```

Comment:
🗺️ `IonParseResult<'a>` has a lifetime. Because this function takes several parameters that have their own lifetimes, we have to explicitly say which lifetime the returned `IonParseResult` is using. In this case, they all have the same lifetime so it doesn't matter.